proto/merge.go

```diff
@@ -0,0 +1,114 @@
+package proto
+
+import (
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/google/go-cmp/cmp"
+	"golang.org/x/exp/slices"
+	protov2 "google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protodesc"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/testing/protocmp"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+// MergedFileDescriptors returns a single FileDescriptorSet containing all the
+// file descriptors registered with both gogoproto and protoregistry. When
+// merging it also performs the following checks:
+// - check that all file descriptors' import paths are correct (i.e. match
+// their fully-qualified package name).
+// - check that if both gogo and protoregistry import the same file descriptor,
+// that these are identical under proto.Equal.
+func MergedFileDescriptors() (*descriptorpb.FileDescriptorSet, error) {
+	fds := &descriptorpb.FileDescriptorSet{}
+
+	// load gogo proto file descriptors
+	gogoFds := AllFileDescriptors()
+	gogoFdsMap := map[string]*descriptorpb.FileDescriptorProto{}
+	for _, compressedBz := range gogoFds {
+		rdr, err := gzip.NewReader(bytes.NewReader(compressedBz))
+		if err != nil {
+			return nil, err
+		}
+
+		bz, err := io.ReadAll(rdr)
+		if err != nil {
+			return nil, err
+		}
+
+		fd := &descriptorpb.FileDescriptorProto{}
+		err = protov2.Unmarshal(bz, fd)
+		if err != nil {
+			return nil, err
+		}
+
+		err = CheckImportPath(*fd.Name, *fd.Package)
+		if err != nil {
+			panic(err)
+		}
+
+		fds.File = append(fds.File, fd)
+		gogoFdsMap[*fd.Name] = fd
+	}
+
+	// load any protoregistry file descriptors not in gogo
+	protoregistry.GlobalFiles.RangeFiles(func(fileDescriptor protoreflect.FileDescriptor) bool {
+		fd := protodesc.ToFileDescriptorProto(fileDescriptor)
+		if err := CheckImportPath(*fd.Name, *fd.Package); err != nil {
+			panic(err)
+		}
+
+		gogoFd, found := gogoFdsMap[fileDescriptor.Path()]
+		// If we already loaded gogo's file descriptor, compare that the 2
+		// are strictly equal.
+		if found {
+			if !protov2.Equal(gogoFd, fd) {
+				diff := cmp.Diff(fd, gogoFd, protocmp.Transform())
+				panic(fmt.Errorf("got different file descriptors for %s; %s", *fd.Name, diff))
+			}
+		} else {
+			fds.File = append(fds.File, protodesc.ToFileDescriptorProto(fileDescriptor))
+		}
+
+		return true
+	})
+
+	slices.SortFunc(fds.File, func(x, y *descriptorpb.FileDescriptorProto) bool {
+		return *x.Name < *y.Name
+	})
+
+	return fds, nil
+}
+
+// MergedRegistry returns a protodesc.Resolver that acts as a single registry
+// which contains all the file descriptors registered with both gogoproto and
+// protoregistry.
+func MergedRegistry() (protodesc.Resolver, error) {
```

Comment:
This returns a `protodesc.Resolver`, which is what `protoregistry.GlobalFiles` implements.

There's also a `protodesc.MessageTypeResolver`, which is what `protoregistry.GlobalTypes` implements (and used in the SDK).

Is there an easy wasy to go from `Files` to `Types`? Theoretically yes, but I didn't find any API in protoreflect stuff. Should I add a `MergedTypes()` function in this file too?

Reply:
No... that would be quite hard

Reply:
We use GlobalTypes in 3 instances in the SDK. What can we do if pulsar files are not registered?

Reply:
Where do we use? We can use dynamicpb

Reply:
https://github.com/cosmos/cosmos-sdk/blob/80dd55f79bba8ab675610019a5764470a3e2fef9/x/tx/textual/any.go#L67
https://github.com/cosmos/cosmos-sdk/blob/80dd55f79bba8ab675610019a5764470a3e2fef9/core/appconfig/config.go#L69

Reply:
For AppConfig I guess we can use the merged registry if we need to support gogo there 

For textual we should allow parameterizing the type resolver and also default to dynamicpb if not found